storage/disk/disk.go

```diff
@@ -235,6 +240,220 @@ func (db *Store) NewTransaction(ctx context.Context, params ...storage.Transacti
 	return newTransaction(xid, write, underlying, context, db.pm, db.partitions, db), nil
 }
 
+// Truncate implements the storage.Store interface. This method must be called within a transaction.
+func (db *Store) Truncate(ctx context.Context, txn storage.Transaction, params storage.TransactionParams, it storage.Iterator) error {
+	var err error
+
+	newDB, backupDir, err := db.backupAndLoadDB()
+	if err != nil {
+		return wrapError(err)
+	}
+
+	// write new bundle policy and data into the new DB
+	underlying := newDB.NewTransaction(true)
+	xid := atomic.AddUint64(&db.xid, uint64(1))
+	underlyingTxn := newTransaction(xid, true, underlying, params.Context, db.pm, db.partitions, nil)
+
+	for {
+		var update *storage.Update
+
+		update, err = it.Next()
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			return wrapError(err)
+		}
+
+		if update.IsPolicy {
+			err = underlyingTxn.UpsertPolicy(ctx, update.Path, update.Value)
+			if err != nil {
+				if err != badger.ErrTxnTooBig {
+					return wrapError(err)
+				}
+
+				_, err = underlyingTxn.Commit(ctx)
```

Comment:
What happens when OPA crashes while in the middle of a truncate operation? Does it restart from the old version of the DB or does it restart w/ the new DB that may have had zero or more write operations committed?

Reply:
If OPA crashes in the middle of a truncate operation, it will start with the data in the old DB. The database switch op which happens after all the data is written is atomic.